File: ios/Talise/Features/Onboarding/SignInScreen.swift
import SwiftUI

/// Continue-with-Google screen used as step 1 of the onboarding flow.
/// Sits over the shared `OnboardingBackground` so the green wash and
/// frosted bloom continue from the Welcome hero into auth and on to
/// the handle / PIN / permissions steps.
///
/// Reuses `ZkLoginCoordinator.shared.signIn()` — does NOT reimplement
/// auth. On success the resulting `UserDTO` is passed up to
/// `OnboardingRoot` so the rest of the onboarding flow can run.
struct SignInScreen: View {
    let onSignedIn: (UserDTO) -> Void
    @State private var signingIn = false
    @State private var error: String?

    /// Letter-spacing helper — same `-size × 0.03` ratio used across
    /// the onboarding flow (matches the Figma "-0.705 ls @ 23.5pt"
    /// headline spec).
    private func kern(_ size: CGFloat) -> CGFloat { -size * 0.03 }

    var body: some View {
        ZStack {
            OnboardingBackground()

            VStack(spacing: 0) {
                // Top spacer — leaves room for the OnboardingRoot
                // progress-bar overlay (mounted by the coordinator
                // for this step).
                Spacer().frame(height: 70)

                Spacer()

                hero
                    .frame(width: 96, height: 96)

                Text("Welcome to Talise")
                    .font(TaliseFont.heading(26, weight: .semibold))
                    .kerning(kern(26))
                    .foregroundStyle(TaliseColor.fg)
                    .padding(.top, 28)

                Text("One Google account. One Sui address.\nNo seed phrase, no setup.")
                    .font(TaliseFont.body(14, weight: .light))
                    .kerning(kern(14))
                    .foregroundStyle(TaliseColor.fgMuted)
                    .multilineTextAlignment(.center)
                    .padding(.top, 10)
                    .padding(.horizontal, 32)

                Spacer()

                if let error {
                    Text(error)
                        .font(TaliseFont.body(12))
                        .kerning(kern(12))
                        .foregroundStyle(TaliseColor.danger)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 24)
                        .padding(.bottom, 8)
                }

                continueWithGoogleButton
                    .padding(.horizontal, 24)

                Text("By continuing you agree to our Terms and Privacy.")
                    .font(TaliseFont.body(11, weight: .light))
                    .kerning(kern(11))
                    .foregroundStyle(TaliseColor.fgDim)
                    .frame(maxWidth: .infinity, alignment: .center)
                    .padding(.top, 14)
                    .padding(.bottom, 28)
            }
        }
    }

    // ── CTA ────────────────────────────────────────────────────────

    /// White capsule CTA matching the Welcome hero "Get Started"
    /// shape (54pt tall, capsule clip, white fill, bg text). The
    /// Google G mark sits inline-left of the title, ~20pt wide.
    private var continueWithGoogleButton: some View {
        Button {
            Task { await beginSignIn() }
        } label: {
            HStack(spacing: 10) {
                if signingIn {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(TaliseColor.bg)
                        .frame(width: 20, height: 20)
                } else {
                    Image("GoogleG")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 20, height: 20)
                }
                Text("Continue with Google")
                    .font(TaliseFont.body(15, weight: .medium))
                    .kerning(kern(15))
                    .foregroundStyle(TaliseColor.bg)
            }
            .frame(maxWidth: .infinity)
            .frame(height: 54)
            .background(TaliseColor.fg)
            .clipShape(Capsule())
        }
        .buttonStyle(.plain)
        .disabled(signingIn)
    }

    // ── Hero (Talise pinwheel) ─────────────────────────────────────

    @ViewBuilder
    private var hero: some View {
        if UIImage(named: "TaliseLogo") != nil {
            Image("TaliseLogo")
                .resizable()
                .scaledToFit()
        } else {
            Canvas { ctx, size in
                let cx = size.width / 2
                let cy = size.height / 2
                let r: CGFloat = size.width * 0.22
                for i in 0..<4 {
                    let angle = CGFloat(i) * .pi / 2
                    var t = CGAffineTransform(translationX: cx, y: cy)
                    t = t.rotated(by: angle)
                    t = t.translatedBy(x: 0, y: -size.height * 0.28)
                    let rect = CGRect(
                        x: -r * 0.45, y: -r * 0.55,
                        width: r * 0.9, height: r * 1.15
                    ).applying(t)
                    ctx.fill(Path(ellipseIn: rect), with: .color(.white))
                }
            }
        }
    }

    private func beginSignIn() async {
        signingIn = true
        error = nil
        defer { signingIn = false }
        do {
            let result = try await ZkLoginCoordinator.shared.signIn()
            onSignedIn(result.user)
        } catch GoogleSignInService.SignInError.cancelled {
            // Quiet — the user explicitly backed out of the OAuth sheet.
        } catch {
            self.error = error.localizedDescription
        }
    }
}

// The leading icon on the "Continue with Google" CTA uses the real
// Google "G" mark from the asset catalog (`Image("GoogleG")`), per
// Google's sign-in branding guidelines.

File: android/app/src/main/java/io/talise/app/feature/onboarding/SignInScreen.kt
package io.talise.app.feature.onboarding

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import io.talise.app.R
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType

/**
 * Sign-in entry, mirrors iOS `Onboarding/SignInScreen`. Native zkLogin via Google:
 * Credential Manager returns a nonce-bound ID token → `/api/auth/mobile/exchange` →
 * bearer. Layout matches iOS: shared onboarding backdrop (mossy-green wash → black with a
 * top-right bloom), Talise pinwheel hero, welcome copy, a white "Continue with Google"
 * capsule CTA, then the beta + legal notes. iOS also offers Apple; Android is Google-only,
 * so the subtitle names Google alone (auth wiring unchanged).
 */
@Composable
fun SignInScreen(vm: SignInViewModel = viewModel()) {
    val state by vm.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val loading = state.loading
    val error = state.error

    // True once at least one successful sign-in happened on this device — drives the
    // "Welcome back" copy for returning users (iOS `talise.hasSignedInBefore`).
    val returningUser = remember {
        OnboardingPrefs.of(context).getBoolean(OnboardingPrefs.KEY_HAS_SIGNED_IN_BEFORE, false)
    }

    Box(Modifier.fillMaxSize().background(TaliseColors.bg)) {
        OnboardingBackground(Modifier.fillMaxSize())

        Column(
            modifier = Modifier.fillMaxSize(),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            // Leaves room for the progress-bar overlay the iOS coordinator mounts here.
            Spacer(Modifier.height(70.dp))
            Spacer(Modifier.weight(1f))

            // Talise pinwheel hero, the real brand mark (96×96, fit).
            Image96Logo()

            Text(
                if (returningUser) "Welcome back" else "Welcome to Talise",
                style = TaliseType.heading(26.sp, FontWeight.SemiBold),
                letterSpacing = (-0.78).sp,
                color = TaliseColors.fg,
                modifier = Modifier.padding(top = 28.dp),
            )
            Text(
                if (returningUser) "Sign in to your Talise account."
                else "One tap with Google.\nNo seed phrase, no setup.",
                style = TaliseType.body(14.sp, FontWeight.Light),
                letterSpacing = (-0.42).sp,
                color = TaliseColors.fgMuted,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 10.dp, start = 32.dp, end = 32.dp),
            )

            Spacer(Modifier.weight(1f))

            if (error != null) {
                Text(
                    error,
                    style = TaliseType.body(12.sp),
                    color = TaliseColors.danger,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(horizontal = 24.dp).padding(bottom = 8.dp),
                )
            }

            // Provider CTA, white 54dp capsule, Google G leading the label.
            ContinueWithGoogleButton(
                loading = loading,
                onClick = { vm.signInWithGoogle(context) },
                modifier = Modifier.padding(horizontal = 24.dp),
            )

            Text(
                "Talise is in private beta. Access is invite-only.",
                style = TaliseType.body(11.sp, FontWeight.Light),
                letterSpacing = (-0.33).sp,
                color = TaliseColors.fgDim,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth().padding(top = 12.dp, start = 24.dp, end = 24.dp),
            )
            Text(
                "By continuing you agree to our Terms and Privacy.",
                style = TaliseType.body(11.sp, FontWeight.Light),
                letterSpacing = (-0.33).sp,
                color = TaliseColors.fgDim,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 14.dp, bottom = 28.dp, start = 24.dp, end = 24.dp),
            )
        }
    }
}

/** Talise pinwheel brand mark, 96×96 fit, iOS `hero` (`Image("TaliseLogo")`). */
@Composable
private fun Image96Logo() {
    androidx.compose.foundation.Image(
        painter = painterResource(R.drawable.taliselogo),
        contentDescription = "Talise",
        contentScale = ContentScale.Fit,
        modifier = Modifier.size(96.dp),
    )
}

/**
 * White capsule CTA, iOS `continueWithGoogleButton` (54dp tall, white fill, dark ink label,
 * 20dp leading Google G). While loading the G is swapped for a dark spinner; label unchanged.
 */
@Composable
private fun ContinueWithGoogleButton(
    loading: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(54.dp)
            .clip(RoundedCornerShape(27.dp))
            .background(TaliseColors.fg)
            .clickable(enabled = !loading) { onClick() },
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (loading) {
            CircularProgressIndicator(
                color = TaliseColors.bg,
                strokeWidth = 2.dp,
                modifier = Modifier.size(20.dp),
            )
        } else {
            androidx.compose.foundation.Image(
                painter = painterResource(R.drawable.googleg),
                contentDescription = null,
                contentScale = ContentScale.Fit,
                modifier = Modifier.size(20.dp),
            )
        }
        Spacer(Modifier.size(10.dp))
        Text(
            "Continue with Google",
            style = TaliseType.body(15.sp, FontWeight.Medium),
            letterSpacing = (-0.45).sp,
            color = TaliseColors.bg,
        )
    }
}

// OnboardingBackground moved to its own file (OnboardingBackground.kt) to mirror
// the iOS file layout now that every onboarding step shares it.

